mlora/model_mixlora.py

```diff
@@ -0,0 +1,638 @@
+
+        # feed forward fully connected
+        score_norm_data = self.ffn_norm_.forward(data)
+        common_w1 = self.w1_.weight_.forward(score_norm_data)
+        common_w3 = self.w3_.weight_.forward(score_norm_data)
+
+        final_ffn_output = None
+        for lora_config in input_args.lora_batch_data_config_:
+            moe_name = lora_config.adapter_name_
+            start_idx = lora_config.batch_start_idx_
+            end_idx = lora_config.batch_end_idx_
+
+            if moe_name == "" or moe_name not in self.moes_:
+                continue
+
+            moe_layer = self.moes_[moe_name]
+
+            norm_data = score_norm_data[start_idx:end_idx]
+            w1_data = common_w1[start_idx:end_idx]
+            w3_data = common_w3[start_idx:end_idx]
+
+            # routing to experts
+            router_logits = moe_layer.gate_.forward(norm_data)
+            routing_weights = F.softmax(
+                router_logits, dim=1, dtype=torch.float)
+            routing_weights, selected_experts = torch.topk(
+                routing_weights, moe_layer.topk_, dim=-1
+            )
+            routing_weights /= routing_weights.sum(dim=-1, keepdim=True)
+
+            # mix of experts
+            final_hidden_states = None
+            for expert_idx in range(moe_layer.experts_):
+                lora_name = f"moe.{moe_name}.experts.{expert_idx}"
+                if lora_name in self.w1_.loras_:
+                    w1 = w1_data + \
+                        self.w1_.loras_[lora_name].forward(norm_data)
+                else:
+                    w1 = w1_data
+
+                if lora_name in self.w3_.loras_:
+                    w3 = w3_data + \
+                        self.w3_.loras_[lora_name].forward(norm_data)
+                else:
+                    w3 = w3_data
+
+                silu_result = F.silu(w1) * w3
+                if lora_name in self.w2_.loras_:
+                    hidden_state = self.w2_.weight_.forward(
+                        silu_result
+                    ) + self.w2_.loras_[lora_name].forward(silu_result)
+                else:
+                    hidden_state = self.w2_.weight_.forward(silu_result)
+
+                expert_mask = selected_experts == expert_idx
+                expert_weights = (routing_weights * expert_mask).sum(
+                    dim=-1, keepdim=True
+                )
+                current_hidden_states = hidden_state.mul_(expert_weights)
+                if final_hidden_states is None:
+                    final_hidden_states = current_hidden_states
+                else:
+                    final_hidden_states.add_(current_hidden_states)
+
+            if final_ffn_output is None:
+                final_ffn_output = final_hidden_states
+            else:
+                final_ffn_output = torch.cat(
+                    [final_ffn_output, final_hidden_states], dim=0
+                )
+
+        return data + final_ffn_output
```

Comment:
If lora is disabled,  `final_ffn_output` will be none type.

Reply:
``` if final_ffn_output is not None:
            data = data + final_ffn_output

        return data```

Reply:
have fixed in https://github.com/TUDB-Labs/multi-lora-fine-tune/pull/136/commits/596d01822ab903f5bf54a7834deaecacf44f1db7